File: server/src/modules/purchases/purchase.model.ts
import mongoose, { Schema, Document } from 'mongoose';

export interface IPurchaseItem {
  productId: mongoose.Types.ObjectId;
  productName: string;
  sku: string;
  quantity: number;
  costPrice: number;
  subtotal: number;
}

export interface IPurchase extends Document {
  poNumber: string;
  supplierId: mongoose.Types.ObjectId;
  items: IPurchaseItem[];
  totalAmount: number;
  status: 'PENDING' | 'PARTIAL' | 'RECEIVED' | 'CANCELLED';
  orderedBy: mongoose.Types.ObjectId;
  receivedAt?: Date;
  createdAt: Date;
  updatedAt: Date;
}

const purchaseItemSchema = new Schema(
  {
    productId: { type: Schema.Types.ObjectId, ref: 'Product', required: true },
    productName: { type: String, required: true },
    sku: { type: String, required: true },
    quantity: { type: Number, required: true, min: 1 },
    costPrice: { type: Number, required: true, min: 0 },
    subtotal: { type: Number, required: true, min: 0 },
  },
  { _id: false }
);

const purchaseSchema = new Schema<IPurchase>(
  {
    poNumber: { type: String, required: true, unique: true },
    supplierId: { type: Schema.Types.ObjectId, ref: 'Supplier', required: true },
    items: [purchaseItemSchema],
    totalAmount: { type: Number, required: true, min: 0 },
    status: {
      type: String,
      enum: ['PENDING', 'PARTIAL', 'RECEIVED', 'CANCELLED'],
      default: 'PENDING',
    },
    orderedBy: { type: Schema.Types.ObjectId, ref: 'User', required: true },
    receivedAt: { type: Date },
  },
  { timestamps: true }
);

export const PurchaseModel = mongoose.model<IPurchase>('Purchase', purchaseSchema);

File: mobile/lib/features/purchases/data/purchase_model.dart
class PurchaseItemModel {
  final String productId;
  final String productName;
  final String sku;
  final int quantity;
  final double costPrice;
  final double subtotal;

  PurchaseItemModel({
    required this.productId,
    required this.productName,
    required this.sku,
    required this.quantity,
    required this.costPrice,
    required this.subtotal,
  });

  factory PurchaseItemModel.fromJson(Map<String, dynamic> json) {
    return PurchaseItemModel(
      productId: json['productId'] ?? '',
      productName: json['productName'] ?? '',
      sku: json['sku'] ?? '',
      quantity: (json['quantity'] as num?)?.toInt() ?? 0,
      costPrice: (json['costPrice'] as num?)?.toDouble() ?? 0.0,
      subtotal: (json['subtotal'] as num?)?.toDouble() ?? 0.0,
    );
  }
}

class PurchaseModel {
  final String id;
  final String poNumber;
  final Map<String, dynamic> supplierId;
  final List<PurchaseItemModel> items;
  final double totalAmount;
  final String status;
  final DateTime? receivedAt;
  final DateTime createdAt;

  PurchaseModel({
    required this.id,
    required this.poNumber,
    required this.supplierId,
    required this.items,
    required this.totalAmount,
    required this.status,
    this.receivedAt,
    required this.createdAt,
  });

  factory PurchaseModel.fromJson(Map<String, dynamic> json) {
    return PurchaseModel(
      id: json['_id'] ?? json['id'] ?? '',
      poNumber: json['poNumber'] ?? '',
      supplierId: json['supplierId'] ?? {},
      items: (json['items'] as List<dynamic>?)
              ?.map((i) => PurchaseItemModel.fromJson(i))
              .toList() ??
          [],
      totalAmount: (json['totalAmount'] as num?)?.toDouble() ?? 0.0,
      status: json['status'] ?? 'PENDING',
      receivedAt: json['receivedAt'] != null ? DateTime.tryParse(json['receivedAt']) : null,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : DateTime.now(),
    );
  }
}
